book/forms.md

```diff
@@ -261,119 +261,93 @@ slow in your browser, so typing into input forms is actually going to
-having the `InputLayout` have a child `BlockLayout`, but I'm skipping
-it here for simplicity.
-
-Ok, next up, button clicks. We need to extend `handle_click` with
-button support. That requires modifying the condition in the big
-`while` loop and then adding a new case to the big `if` statement:
-
-``` {.python indent=16}
-# ...
-elif elt.tag == "button":
-    self.submit_form(elt)
-# ...
-```
-
-Third, we need to find the form containing our button. That can happen
-inside `submit_form`:[^3]
+Once we've found the button, we need to find the form that it's in, by
+walking up the HTML tree:[^3]
 
 [^3]: Fun fact: HTML standardizes the `form` attribute for _input
     elements_, which in principle allows an input element to be
     outside the form it is supposed to be submitted with. But no
     browser implements that.
 
-``` {.python}
-def submit_form(self, elt):
-    while elt and elt.tag != "form":
+``` {.python indent=12}
+elif elt.tag == "button":
+    while elt:
+        if elt.tag == "form" and "action" in elt.attributes:
+            return self.submit_form(elt)
         elt = elt.parent
-    if not elt: return
 ```
 
-Fourth, we need to find all of the input elements inside this form:
+The `submit_form` method is then in charge of finding all of the input
+elements, encoding them in the right way, and sending the `POST`
+request. First, we look through all the descendents of the `form` to
+find `input` elements:
 
 ``` {.python}
-def find_inputs(elt, out):
-    if not isinstance(elt, ElementNode): return
-    if elt.tag == "input" and "name" in elt.attributes:
-        out.append(elt)
-    for child in elt.children:
-        find_inputs(child, out)
-    return out
+class Tab:
+    def submit_form(self, elt):
+        inputs = [node for node in tree_to_list(elt)
+                  if isinstance(node, Element)
+                  and node.tag == "input"
+                  and "name" in node.attributes]
 ```
 
-Fifth, we can form-encode the resulting parameters:
-
+For each of those `input` elements, we need to extract the `name`
+attribute and the `value` attribute, and _form-encode_ both of them.
+Form encoding is how the name/value pairs are formatted in the HTTP
+`POST` request. Basically: name, then equal sign, then value; and
+name-value pairs are separated by ampersands:
 
 ``` {.python}
-def submit_form(self, elt):
+class Tab:
+    def submit_form(self, elt):
+        # ...
+        body = ""
+        for input in inputs:
+            name = input.attributes["name"]
+            value = input.attributes.get("value", "")
+            body += "&" + name + "=" + value
+        body = body [1:]
+```
+
+Now, any time you see something like this, you've got to ask: what if
+the name or the value has an equal sign or an ampersand in it? So form
+encoding has special handling for special characters:
+
+``` {.python indent=8}
+for input in inputs:
+    # ...
+    name = percent_encode(name)
```

Comment:
Also, change this to the python method that does it?

Reply:
Done